### PR TITLE
feat(core): ship some types

### PR DESCRIPTION
**Pull request by boneskull**
This adds _some_ types to `lavamoat-core`.

Interesting stuff includes the `LavaMoatPolicy` type and the types for `endowmentsToolkit.js`

### CODE REVIEW

**Comment by boneskull** [comment]
* **#732** <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/732?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 3  dependent PRs ([#733](https://github.com/LavaMoat/LavaMoat/pull/733) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/733?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#752](https://github.com/LavaMoat/LavaMoat/pull/752) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/752?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#755](https://github.com/LavaMoat/LavaMoat/pull/755) <a href="https://app.graphite.dev/github/pr/LavaMoat/LavaMoat/755?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @boneskull and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by boneskull** [comment]
This includes some fixes to the eslint configuration, as it's presently confused about whether or not it should lint generated types.  It shouldn't, and it doesn't in CI, but it does locally.

File: packages/core/package.json [hunk 1]
```diff
@@ -12,9 +12,11 @@
   "dependencies": {
     "json-stable-stringify": "^1.0.2",
     "lavamoat-tofu": "^6.2.1",
-    "merge-deep": "^3.0.3"
+    "merge-deep": "^3.0.3",
+    "type-fest": "^4.3.3"
```
Comment (legobeat) [inline packages/core/package.json]: Can we achieve this without needing a new runtime dependency? Seems like this belongs in devDependencies.

File: packages/core/src/endowmentsToolkit.js [hunk 2]
```diff
@@ -328,12 +418,19 @@ function endowmentsToolkit({
...
+   * @returns {{prop: PropertyDescriptor|null, receiver: object|null}}
+   */
   function getPropertyDescriptorDeep(target, key) {
+    /** @type {object|null} */
     let receiver = target
     // eslint-disable-next-line no-constant-condition
     while (true) {
       // abort if this is the end of the prototype chain.
-      if (!receiver) {
+      if (receiver === null) {
```
Comment (legobeat) [inline packages/core/src/endowmentsToolkit.js]: Potential behavior change, I think? Can those be broken out as separate PR(s) and merged independently to shipping types?

File: packages/core/package.json (hunk 1 above)
Comment (boneskull) [reply packages/core/package.json]: It's not a dev dep because we export a type that uses it (policy schema).

File: packages/core/src/endowmentsToolkit.js (hunk 2 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js]: I've reverted this

File: packages/core/src/endowmentsToolkit.js [hunk 3]
```diff
@@ -1,12 +1,26 @@
-// the contents of this file will be copied into the prelude template
-// this module has been written so that it required directly or copied and added to the template with a small wrapper
```
Comment (naugtur) [inline packages/core/src/endowmentsToolkit.js:2]: We should keep this comment, unless it's no longer true.

File: packages/core/src/generatePolicy.js [hunk 4]
```diff
@@ -126,7 +170,7 @@ function createModuleInspector(opts = {}) {
         allowReturnOutsideFunction: true,
         errorRecovery: true,
       })
-    if (includeDebugInfo && ast.errors && ast.errors.length) {
+    if (includeDebugInfo && isParsedAST(ast) && ast.errors.length) {
```
Comment (naugtur) [inline packages/core/src/generatePolicy.js:173]: This is clever. And I don't even mind the runtime overhead vs clarity in this case.

File: packages/core/src/generatePolicy.js [hunk 5]
```diff
@@ -246,16 +316,28 @@ function createModuleInspector(opts = {}) {
...
+    packageBuiltins = [...(packageBuiltins ?? []), ...moduleBuiltins]
     packageToBuiltinImports.set(packageName, packageBuiltins)
   }
 
-  function generatePolicy({ policyOverride, includeDebugInfo } = {}) {
+  /**
+   * @type {GeneratePolicyFn}
+   */
+  function generatePolicy({ policyOverride, includeDebugInfo = false }) {
+    /** @type {import('./schema').Resources} */
```
Comment (naugtur) [inline packages/core/src/generatePolicy.js:327]: we're in a closure so if I understood correctly, you can do a typedef import and get ResourcePolicy etc. directly without reexporting it. Was that the case that only top level typedef needs to be avoided because it's reexported?

Comment (boneskull) [reply packages/core/src/generatePolicy.js:327]: Yes, it could be used here without worrying about an export (I think!), but if `Resources` becomes generic in the future, we would need to re-implement those  type args in `@template` tags.  That said, you could probably get away with it here--kind of like you can get away with writing JS without semicolons.

_Personally,_ I don't want to think too hard about whether or not it's safe to create the `@typedef` reference, so I choose not to and make that tradeoff.  Similarly, I don't like thinking about whether or not ASI is going to work either, so I always use semicolons. :smile:

File: packages/core/src/generatePolicy.js (hunk 4 above)
Comment (boneskull) [reply packages/core/src/generatePolicy.js:173]: `ast.errors` doesn't mean very much to TS' type inference, but a proper type guard like `isParsedAST()` does.

File: packages/core/src/endowmentsToolkit.js (hunk 3 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js:2]: it's there; it's just moved.

File: packages/core/src/endowmentsToolkit.js [hunk 6]
```diff
@@ -379,7 +484,7 @@ function endowmentsToolkit({
     endowmentSources.forEach((source) => {
       const descriptors = Object.getOwnPropertyDescriptors(source)
       Object.values(descriptors).forEach((desc) => {
-        if ('get' in desc) {
+        if ('get' in desc && desc.get) {
```
Comment (kumavis) [inline packages/core/src/endowmentsToolkit.js:495]: I usually do `desc.get !== undefined`

File: packages/core/src/generatePolicy.js [hunk 7]
```diff
@@ -160,6 +213,7 @@ function createModuleInspector(opts = {}) {
     if (includeDebugInfo) {
       const moduleDebug = debugInfo[moduleRecord.specifier]
       moduleDebug.sesCompat = {
+        // XXX: I don't think this is needed, since it appears we overwrite all properties
```
Comment (kumavis) [inline packages/core/src/generatePolicy.js]: this is just for the debug file. what do you mean about overwriting all props?

File: packages/core/src/endowmentsToolkit.js (hunk 6 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js:495]: i want it to be truthy

File: packages/core/src/generatePolicy.js (hunk 7 above)
Comment (boneskull) [reply packages/core/src/generatePolicy.js]: `compatWarnings` has props `primordialMutations`, `strictModeViolations` and `dynamicRequires`, all of which are overwritten below.  I don't think there are other props.

Comment (boneskull) [reply packages/core/src/generatePolicy.js]: (meaning that `...compatWarnings` is unnecessary. at least, I think)

File: packages/core/src/endowmentsToolkit.js (hunk 6 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js:495]: fwiw `'foo' in bar` is the TypeScript-Approved:tm: way to check if a prop "has" something; skipping right to `foo.bar` will typically be an error if it's unclear from type inference

Comment (naugtur) [reply packages/core/src/endowmentsToolkit.js:495]: in the case of a descriptor there's no difference between it being truthy or not-undefined because it can only be a function or missing. 
Is this for typescript only?

File: packages/core/src/generatePolicy.js (hunk 5 above)
Comment (naugtur) [reply packages/core/src/generatePolicy.js:327]: As the best practices around this are far from set in stone, I'm trying to explore options in case we come up with something that's more convenient. 

Also note that making this pattern more popular makes SES slower, because it relies on a transform to replace each `import(` with something else

Comment (boneskull) [reply packages/core/src/generatePolicy.js:327]: The transform in question doesn't happen in SES, fwiw.  It means we have to apply it ourselves; hence https://github.com/endojs/endo/pull/1812

File: packages/core/src/endowmentsToolkit.js (hunk 6 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js:495]: According to TS, `get` can be `null`.  And `get: null` would be true for `'get' in desc`; hence the truthiness check.  TS' ECMAScript libs aren't the spec, but I trust them.

**Comment by boneskull** [comment]
I'm going to split this up into two or three separate PRs.

File: packages/core/src/endowmentsToolkit.js (hunk 6 above)
Comment (naugtur) [reply packages/core/src/endowmentsToolkit.js:495]: If we want to support the folks using typescript directly on platforms we don't yet foresee (poor souls) this seems like the right thing to do.

**Comment by naugtur** [comment]
> I'm going to split this up into two or three separate PRs.

I was ready to say we approve, merge and see what happens, but a split might serve us well. That is, if your tooling can figure out the tree of dependent items.

File: packages/core/src/endowmentsToolkit.js [hunk 8]
```diff
@@ -263,20 +329,29 @@ function endowmentsToolkit({
...
+      const getter = sourcePropDesc.get
       wrappedPropDesc.get = function () {
+        // eslint-disable-next-line @typescript-eslint/no-this-alias
         const receiver = this
         // replace the "receiver" value if it points to fake parent
         const receiverRef =
           receiver === unwrapFromGlobalThis ? unwrapToGlobalThis : receiver
         // sometimes getters replace themselves with static properties, as seen wih the FireFox runtime
-        const result = Reflect.apply(sourcePropDesc.get, receiverRef, [])
+        const result = Reflect.apply(getter, receiverRef, [])
```
Comment (naugtur) [inline packages/core/src/endowmentsToolkit.js]: (must) Let's make sure this is not breaking the code. I'm pretty sure the comment above is explaining why we're not using a local copy but referencing the getter from the original source and this change would break it.

**Review by naugtur — CHANGES_REQUESTED** [review]
I've spotted a new bug being introduced, I think: 
https://github.com/LavaMoat/LavaMoat/pull/732/files#diff-5ed0581fdb5b3ec8105f7f2e6773d6f5db45dbe74dccf2a921cc9c0e40ecda60R354

File: packages/core/src/endowmentsToolkit.js (hunk 8 above)
Comment (legobeat) [reply packages/core/src/endowmentsToolkit.js]: Regardless of outcome:  This seems like a non-type-related refactor that could/should be broken out as separate change? Or maybe I'm missing some context on why this change is here?

Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js]: Sorry this wasn't explained; TS' type inference doesn't know that `sourcePropDesc.get` is not `undefined` at this point, because it doesn't know when `wrappedPropDesc.get()` is called, so it cannot be sure, despite the conditional on L345 (or L346, depending).

I've changed it to just use a type assertion instead.

To be clear, I didn't make any code changes in this PR that _weren't_ prompted by type checking.

Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js]: By leaving it as-is, we're saying: _`sourcePropDesc.get` is allowed to be `undefined` here, and if so, allow `Reflect.apply()` to throw an exception_.  Just want to confirm that's the intent.

File: packages/core/src/endowmentsToolkit.js [hunk 9]
```diff
@@ -295,17 +379,26 @@ function endowmentsToolkit({
...
     if (sourcePropDesc.set) {
+      const setter = sourcePropDesc.set
       wrappedPropDesc.set = function (value) {
         // replace the "receiver" value if it points to fake parent
+        // eslint-disable-next-line @typescript-eslint/no-this-alias
         const receiver = this
         const receiverRef =
           receiver === unwrapFromGlobalThis ? unwrapToGlobalThis : receiver
-        return Reflect.apply(sourcePropDesc.set, receiverRef, [value])
+        return Reflect.apply(setter, receiverRef, [value])
```
Comment (naugtur) [inline packages/core/src/endowmentsToolkit.js]: (must) same issue as with https://github.com/LavaMoat/LavaMoat/pull/732/files#diff-5ed0581fdb5b3ec8105f7f2e6773d6f5db45dbe74dccf2a921cc9c0e40ecda60R354

Sorry I didn't point it out here. I did in a different PR that was showing it. 

GH will not let me add a suggestion that fixes it...

**Comment by naugtur** [comment]
BTW, trees managed by Graphite are cool, but they encourage amending the existing commit with review fixes and it makes it harder for the reviewer to see what changes were made since they last saw it. 
Still net positive, but I thought I'd point it out ;) 
Back to reviewing the whole thing top to bottom.

File: packages/core/src/endowmentsToolkit.js (hunk 8 above)
Comment (naugtur) [reply packages/core/src/endowmentsToolkit.js]: We could doublecheck before applying. The breaking change was that we cached the getter instead of using the fresh one that could have been updated.

**Comment by boneskull** [comment]
@naugtur No, that's just me squashing commits.  nothing to do with graphite

File: packages/core/src/endowmentsToolkit.js (hunk 9 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js]: done

File: packages/core/src/endowmentsToolkit.js (hunk 8 above)
Comment (boneskull) [reply packages/core/src/endowmentsToolkit.js]: I'm not sure what sort of action would be reasonable here _other_ than just letting it throw.  In other words, I don't know how to improve on that.

**Comment by naugtur** [comment]
> @naugtur No, that's just me squashing commits.  nothing to do with graphite

Ok, in that case, I request that whenever possible/convenient commits remain visible for easier review and you squash them later.

**Comment by boneskull** [comment]
@naugtur Will do.  Ball still in yr court though

**Review by naugtur — APPROVED** [review]
